oscore-cf/src/main/java/org/eclipse/californium/oscore/OSCoreCtx.java

```diff
@@ -0,0 +1,673 @@
+			throw new OSException("Sequence number wrapped");
+		}
+		sender_seq++;
+	}
+
+	/**
+	 * Checks and sets the sequence number for incoming messages.
+	 * 
+	 * @param seq the incoming sequence number
+	 * 
+	 * @throws OSException if the sequence number wraps or if for a replay
+	 */
+	public synchronized void checkIncomingSeq(int seq) throws OSException {
+		if (seq >= seqMax) {
+			LOGGER.severe("Sequence number wrapped, get new OSCore context");
+			throw new OSException(OSExceptionHandler.REPLAY_DETECT);
+		}
+		rollback_recipient_seq = recipient_seq;
+		rollback_recipient_replay = recipient_replay_window;
+		if (seq > recipient_seq) {
+			// Update the replay window
+			int shift = seq - recipient_seq;
+			recipient_replay_window = recipient_replay_window << shift;
+			recipient_seq = seq;
+		} else if (seq == recipient_seq) {
+			throw new OSException(OSExceptionHandler.REPLAY_DETECT);
+		} else { // seq < recipient_seq
+			if (seq + recipient_replay_window_size < recipient_seq) {
+				LOGGER.severe("Message too old");
+				throw new OSException(OSExceptionHandler.REPLAY_DETECT);
+			}
+			// seq+replay_window_size > recipient_seq
+			int shift = this.recipient_seq - seq;
+			int pattern = 1 << shift;
+			int verifier = recipient_replay_window & pattern;
+			verifier = verifier >> shift;
+			if (verifier == 1) {
+				throw new OSException(OSExceptionHandler.REPLAY_DETECT);
+			}
+			recipient_replay_window = recipient_replay_window | pattern;
+		}
+	}
+
+	/**
+	 * Rolls back the latest recipient sequence number update if any
+	 */
+	public synchronized void rollBack() {
+		if (rollback_recipient_replay != -1) {
+			recipient_replay_window = rollback_recipient_replay;
+			rollback_recipient_replay = -1;
+		}
+		if (rollback_recipient_seq != -1) {
+			recipient_seq = rollback_recipient_seq;
+			rollback_recipient_seq = -1;
+		}
+		if (this.rollback_last_block_tag != null) {
+			this.last_block_tag = this.rollback_last_block_tag;
+			this.rollback_last_block_tag = null;
+		}
+	}
+
+	/**
+	 * Get IV length in bytes.
+	 */
+	private static int ivLength(AlgorithmID alg) {
+		switch (alg) {
+		case AES_CCM_16_64_128:
+		case AES_CCM_16_64_256:
+		case AES_CCM_16_128_128:
+		case AES_CCM_16_128_256:
+			return 13;
+		case AES_CCM_64_64_128:
+		case AES_CCM_64_64_256:
+		case AES_CCM_64_128_256:
+		case AES_CCM_64_128_128:
+			return 7;
+		case AES_GCM_128:
+		case AES_GCM_192:
+		case AES_GCM_256:
+			return 12;
+		default:
+			return -1;
+		}
+	}
+
+	final protected static char[] hexArray = "0123456789ABCDEF".toCharArray();
+
+	/**
+	 * Debug function to convert a byte array to a hex string.
+	 * 
+	 * @param bytes the byte array
+	 * 
+	 * @return the hex string
+	 */
+	public static String bytesToHex(byte[] bytes) {
+		String hex = "";
+		for (int j = 0; j < bytes.length; j++) {
+			int v = bytes[j] & 0xFF;
+			hex = hex + "0x" + hexArray[v >>> 4] + hexArray[v & 0x0F] + " ";
```

Comment:
Consider to use `StringBuilder` instead of repeated String concatenation.

Reply:
Or event better, try to re-use org.eclipse.californium.elements.util.StringUtil's methods.

Reply:
I will look into them!